text/0037-anr-rates.md

```diff
@@ -0,0 +1,107 @@
+
+# Motivation
+
+ANR rate is an app health diagnostic (much like crash rate) that Google Play provides to allow developers to compare stability of releases. Google Play Console (platform that Google provides for developers to monitor the performance of their app, where they would find ANR rate) is not accessible to all developers on a project/team. We want to surface ANR rates (and potentially user-perceived ANR rates) in Sentry to consolidate these app health diagnostics in one place.
+
+# Background
+
+ANRs are triggered if the UI thread on an Android app is blocked for too long. Android vitals provides the following [definition](https://developer.android.com/topic/performance/vitals/anr#android-vitals) for ANRs:
+
+**ANR rate:** The percentage of your daily active users who experienced any type of ANR.
+
+**User-perceived ANR rate:** The percentage of your daily active users who experienced at least one *user-perceived ANR*. Currently only ANRs of type `Input dispatching timed out` are considered user-perceived.
+
+**Daily active user**: Unique user who uses your app on a single day on a single device, potentially over multiple sessions. If a user uses your app on more than one device in a single day, each device will contribute to the number of active users for that day. If multiple users use the same device in a single day, this is counted as one active user.
+
+<aside>
+💡 User-perceived ANR rate is a *core vital* meaning that it affects the discoverability of your app on Google Play.
+</aside>
+
+
+ANRs are currently surfaced as error events with tag `mechanism:ANR`. With the data we already have on hand, we can calculate ANR rate as follows:
+
+```bash
+(unique users who experienced error events with tag mechanism: ANR)
+___________________________________________________________________ x 100 
+                    (total unique users)
+```
+
+There are a couple problems with this:
+
+1. Total ANRs is affected by client side sampling and dropped events if org/project is close to error quota
+2. The most accurate count of total unique users will probably come from the sessions dataset, or maybe even the metrics dataset (but again this one is affected by client-side sampling)
+3. Getting this information will require clickhouse queries to two different datasets and post-processing to calculate the ANR rate
+
+Issues outlined in 1 & 2 will result in us showing *wrong* ANR rates and 3 will limit the capabilities of ANR rate - can’t allow sort and search, can’t add it to dashboards or alerts and is not future-proof in case we want to use ANR rates in issue detection.
+
+# Supporting Data
+
+
+# Options Considered
+
+## Introduce a new tag in the sessions dataset (Option 1)
+
+Introduce a new optional tag `freeze_mechanism` in the sessions dataset and track ANRs in sessions in addition to sending the ANR error events. 
```

Comment:
this seems more straightforward to me. we already send `errored` session updates afaik when the application errored, so we could just add a `mechanism` tag to it (and restrict its cardinality to the values we actually care about)

(not sure if we actually do this in all sdks already but we do have affordances for it in the protocol)

Reply:
Agreed that this is the more straightforward solution. I like the idea of attaching mechanism instead restrict cardinality cc: @romtsn